demos/common/hfsm/az_hfsm.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file az_hfsm.h
+ * @brief Hierarchical Finite State Machine implementation.
+ */
+
+#ifndef _az_HFSM_H
+#define _az_HFSM_H
+
+#include <stdint.h>
+
+/**
+ * @brief hfsm event type.
+ *
+ */
+typedef enum
+{
+  AZ_HFSM_ENTRY = 1,
+  AZ_HFSM_EXIT = 2,
+  AZ_HFSM_ERROR = 3,
+  AZ_HFSM_TIMEOUT = 4,
+  AZ_HFSM_EVENT_BASE = 10,
+} az_hfsm_event_type;
+
+#define AZ_HFSM_EVENT(id) ((int32_t)(AZ_HFSM_EVENT_BASE + id))
+
+struct az_hfsm_event
+{
+  az_hfsm_event_type type;
+  void* data;
+};
+
+typedef struct az_hfsm_event az_hfsm_event;
```

Comment:
nit: later you have anonymous structs typedef'd. If it doesn't need a forward declaration should this just be 
```c
typedef struct
{
} az_hfsm_event;